# Corona Virus Prediction using Decision Trees
Use Random Forest / Gradient Boosted Regression Tree / Regression Trees algorithms to perform the following:
1. Predict the chances of catching corona virus by using the provided corona virus dataset.

## Step 1: Import all necessary libraries

In [1]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt

## Step 2: Load the Dataset

In [2]:
coronaData = pd.read_csv("coronavirusdataset.csv")
coronaData.info()
coronaData.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7294 entries, 0 to 7293
Data columns (total 45 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   batch_date                     7294 non-null   object 
 1   test_name                      7294 non-null   object 
 2   swab_type                      7294 non-null   object 
 3   covid19_test_results           7294 non-null   object 
 4   age                            7294 non-null   int64  
 5   high_risk_exposure_occupation  7294 non-null   bool   
 6   high_risk_interactions         2727 non-null   object 
 7   diabetes                       7294 non-null   bool   
 8   chd                            7294 non-null   bool   
 9   htn                            7294 non-null   bool   
 10  cancer                         7294 non-null   bool   
 11  asthma                         7294 non-null   bool   
 12  copd                           7294 non-null   b

batch_date                     test_name swab_type covid19_test_results  \
0  2020-10-20               SARS-CoV-2, NAA     Nasal             Negative   
1  2020-10-20  COVID-19 PCR External Result     Nasal             Negative   
2  2020-10-20       Rapid COVID-19 PCR Test     Nasal             Negative   
3  2020-10-20       Rapid COVID-19 PCR Test     Nasal             Negative   
4  2020-10-20       Rapid COVID-19 PCR Test     Nasal             Negative   

   age  high_risk_exposure_occupation high_risk_interactions  diabetes    chd  \
0   39                          False                    NaN     False  False   
1   56                          False                    NaN     False  False   
2   35                          False                    NaN     False  False   
3   37                          False                    NaN     False  False   
4   42                          False                    NaN     False  False   

     htn  ...  headache  loss_of_smell  loss_of_taste  runny_nose  \
0  False  ...     False          False          False       False   
1  False  ...     False          False          False       False   
2  False  ...     False          False          False       False   
3  False  ...     False          False          False       False   
4  False  ...     False          False          False       False   

   muscle_sore  sore_throat  cxr_findings  cxr_impression  cxr_label  cxr_link  
0        False        False           NaN             NaN        NaN       NaN  
1        False        False           NaN             NaN        NaN       NaN  
2        False        False           NaN             NaN        NaN       NaN  
3        False        False           NaN             NaN        NaN       NaN  
4        False        False           NaN             NaN        NaN       NaN  

[5 rows x 45 columns]

## Step 3: Preprocess the Dataset
Drop the unnecessary columns.Specify the list of unnecessary columns to drop

In [3]:
columns_to_drop = ["batch_date","high_risk_interactions", "temperature", "pulse", "sys", "dia", "rr", "sats", "rapid_flu_results", "rapid_strep_results", "ctab", "labored_respiration", "rhonchi", "wheezes", "days_since_symptom_onset", "cough_severity", "sob_severity", "cxr_findings", "cxr_impression", "cxr_label", "cxr_link"]
coronaData = coronaData.drop(columns=columns_to_drop)

In [4]:
coronaData.head()

test_name swab_type covid19_test_results  age  \
0               SARS-CoV-2, NAA     Nasal             Negative   39   
1  COVID-19 PCR External Result     Nasal             Negative   56   
2       Rapid COVID-19 PCR Test     Nasal             Negative   35   
3       Rapid COVID-19 PCR Test     Nasal             Negative   37   
4       Rapid COVID-19 PCR Test     Nasal             Negative   42   

   high_risk_exposure_occupation  diabetes    chd    htn  cancer  asthma  ...  \
0                          False     False  False  False   False   False  ...   
1                          False     False  False  False   False   False  ...   
2                          False     False  False  False   False   False  ...   
3                          False     False  False  False   False   False  ...   
4                          False     False  False  False   False   False  ...   

   fever    sob  diarrhea  fatigue headache  loss_of_smell  loss_of_taste  \
0    NaN  False     False    False    False          False          False   
1    NaN  False     False    False    False          False          False   
2    NaN  False     False    False    False          False          False   
3    NaN  False     False    False    False          False          False   
4    NaN  False     False    False    False          False          False   

   runny_nose  muscle_sore  sore_throat  
0       False        False        False  
1       False        False        False  
2       False        False        False  
3       False        False        False  
4       False        False        False  

[5 rows x 24 columns]

In [5]:
# List of columns with categorical variables that need to be label encoded
categorical_columns = ["test_name", "swab_type", "covid19_test_results", "fever"]

# Perform label encoding for categorical columns
label_encoder = LabelEncoder()
for col in categorical_columns:
    coronaData[col] = label_encoder.fit_transform(coronaData[col].astype(str))

## Step 4: Split the Dataset

In [6]:
# Separate the target variable from the features
X = coronaData.drop('covid19_test_results', axis=1)  # Features (excluding the target variable)
y = coronaData['covid19_test_results']  # Target variable

# Split the dataset into training and testing sets (80% training, 20% testing)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## Step 5: Train the GBRT Model

In [7]:
model = GradientBoostingClassifier(random_state=42)
model.fit(X_train, y_train)

GradientBoostingClassifier(random_state=42)

## Step 6: Evaluate the model

In [8]:
# Make predictions on the test set
y_pred = model.predict(X_test)

# Print the first few predictions
print("Predictions:")
print(y_pred[:10])

# Evaluate the model
accuracy_training = model.score(X_train, y_train)
accuracy_testing = model.score(X_test, y_test)

print()

# Print evaluation metrics
print("Final test set predictions:", y_pred)
print("Final train set accuracy:", accuracy_training)
print("Final test set accuracy:", accuracy_testing)

Predictions:
[0 0 0 0 0 0 0 0 0 0]

Final test set predictions: [0 0 0 ... 0 0 0]
Final train set accuracy: 0.9972579263067695
Final test set accuracy: 0.9972583961617546


Based on the provided results, it appears that the Gradient Boosted Regression Trees (GBRT) model performed very well on both the training and test sets. Here are the findings based on the output:

- Predictions:
The first line of predictions [0 0 0 0 0 0 0 0 0 0] shows the predicted classes for the first 10 samples in the test set. Each entry corresponds to a single prediction, where 0 indicates a predicted class of "Negative" (meaning not catching the coronavirus) and 1 indicates a predicted class of "Positive" (meaning catching the coronavirus). In this case, all the predictions are 0, indicating that the model predicted that none of the individuals in the first 10 samples would catch the coronavirus.

- Final test set predictions:
The line Final test set predictions: [0 0 0 ... 0 0 0] shows the predicted classes for the entire test set. Like before, all the predictions are 0, suggesting that the model predicted "Negative" for all the test samples.

- Final train set accuracy:
The value 0.9972579263067695 represents the accuracy of the model on the training set. An accuracy of approximately 99.73% indicates that the model performed very well on the training data and was able to correctly classify the vast majority of the training samples.

- Final test set accuracy:
The value 0.9972583961617546 represents the accuracy of the model on the test set. An accuracy of approximately 99.73% on the test data indicates that the model's performance on unseen data (test data) is almost as good as its performance on the training data. This suggests that the model generalizes well to new, unseen samples.

Overall, the high accuracy on both the training and test sets indicates that the GBRT model has learned to classify the coronavirus cases effectively based on the provided features.